File: src/csharp/src/DesignPatterns.Examples/Behavioral/Strategy/PaymentProcessing.cs
namespace DesignPatterns.Examples.Behavioral.Strategy;

public interface IPaymentGateway
{
    string ProcessPayment(decimal amount);
}

public sealed class GatewayAlpha : IPaymentGateway
{
    public string ProcessPayment(decimal amount) =>
        FormattableString.Invariant($"ALPHA:{amount:F2}");
}

public sealed class GatewayBeta : IPaymentGateway
{
    public string ProcessPayment(decimal amount) =>
        FormattableString.Invariant($"BETA:{amount:F2}");
}

public sealed class PaymentCheckout
{
    private IPaymentGateway _gateway;

    public PaymentCheckout(IPaymentGateway gateway) => _gateway = gateway;

    public void SetGateway(IPaymentGateway gateway) => _gateway = gateway;

    public string Charge(decimal amount) => _gateway.ProcessPayment(amount);
}

File: src/python/src/designpatterns_examples/behavioral/strategy/payment_processing.py
"""Checkout com gateway de pagamento plugável (Strategy)."""

from __future__ import annotations

from abc import ABC, abstractmethod


class PaymentGateway(ABC):
    @abstractmethod
    def process_payment(self, amount: float) -> str:
        """Processa o valor e devolve um identificador de resultado determinístico para testes."""


class GatewayAlpha(PaymentGateway):
    def process_payment(self, amount: float) -> str:
        return f"ALPHA:{amount:.2f}"


class GatewayBeta(PaymentGateway):
    def process_payment(self, amount: float) -> str:
        return f"BETA:{amount:.2f}"


class PaymentCheckout:
    def __init__(self, gateway: PaymentGateway) -> None:
        self._gateway = gateway

    def set_gateway(self, gateway: PaymentGateway) -> None:
        self._gateway = gateway

    def charge(self, amount: float) -> str:
        return self._gateway.process_payment(amount)
